src/com/sheepit/client/hardware/gpu/GPUDevice.java

```
@@ -80,8 +82,66 @@ public void setOldId(String id) {
 		this.oldId = id;
 	}
 	
+	public int getRenderbucketSize() {
+		return this.renderbucketSize;
+	}
+	
+	public void setRenderbucketSize(int proposedRenderbucketSize) {
+		int renderbucketSize = getRecommandedRenderbucketSize();    // minimum recommended renderbucket size for GPUs
+		
+		if (proposedRenderbucketSize >= 32) {
+			if (type.equals("CUDA")) {
+				if (proposedRenderbucketSize <= getMaximumTileSize()) {
+					renderbucketSize = proposedRenderbucketSize;
+				}
+				else {
+					renderbucketSize = getRecommandedRenderbucketSize();
+				}
+			}
+			else if (type.equals("OPENCL")) {
+				if (proposedRenderbucketSize <= getMaximumTileSize()) {
+					renderbucketSize = proposedRenderbucketSize;
+				}
+				else {
+					renderbucketSize = getRecommandedRenderbucketSize();
+				}
+			}
+		}
+		
+		this.renderbucketSize = renderbucketSize;
+	}
+	
+	public int getRecommandedRenderbucketSize() {
+		if (type.equals("CUDA")) {
+			// Optimal CUDA-based GPUs Renderbucket algorithm
+			return (getMemory() > 1073741824L) ? 256 : 128;
+		}
+		else if (type.equals("OPENCL")) {
+			// Optimal OpenCL-based GPUs Renderbucket algorithm
+			return (memory > 1073741824L) ? 256 : 128;
+		}
+		else {
+			// This branch should not be reached, but if it does, then set the size to 32x32 pixels (safest option)
+			return 32;
```

Why not throw an IllegalStateException or something similar instead of silently failing?

---

I disagree when it comes to throwing (i'd rather log it), but either way, it shouldn't just silently do something unexpected.

---

Because is not an exception, in my opinion, and neither a failed exit. 
The flow of the routine is: if the underlying technology is CUDA (NVIDIA) based tech, then do this. If it's OPENCL, then do that. If it's a new technology, not CUDA, not OPENCL then don't necessarily fail and move into the safest space (32x32 pixel tile size).
I do agree, however, in the logging comment, but there is a large refactoring to improve the overall log management happening at the same time and we will fix that part from the other project.

---

Well, but if there is going to be a new device type, we have to make changes to the client anyway.

I'm more a fan of failing fast and loudly.
A year or two after we merged this and intel maybe released their new gpus someone might have to debug why `getRecommandedRenderbucketSize` always returns `32` :)

---

I agree it should be explainned a bit more, that's why I'm recommending to push in the opencl/nvidia class BUT the client should NEVER fail.
If it walks on 1 leg instead of 2, it's much better then not walking at all.
Uptime, it really important because people are launching the client and walk away for days.

---

But the client will only fail on startup if I understand the code right.
I.e. it would fail before any work has been done.